sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparisonSuite.scala

```diff
@@ -0,0 +1,141 @@
+    assertEquivalent('a >= v.toInt, 'a.attr.trueIfNotNull)
+    assertEquivalent('a === v.toInt, 'a === v)
+    assertEquivalent('a <=> v.toInt, 'a === v)
+    assertEquivalent('a <= v.toInt, 'a === v)
+    assertEquivalent('a < v.toInt, 'a.attr.falseIfNotNull)
+  }
+
+  test("unwrap casts when literal < min") {
+    val v: Int = negativeInt
+    assertEquivalent('a > v, 'a.attr.trueIfNotNull)
+    assertEquivalent('a >= v, 'a.attr.trueIfNotNull)
+    assertEquivalent('a === v, 'a.attr.falseIfNotNull)
+    assertEquivalent('a <=> v, false)
+    assertEquivalent('a <= v, 'a.attr.falseIfNotNull)
+    assertEquivalent('a < v, 'a.attr.falseIfNotNull)
+  }
+
+  test("unwrap casts when literal is within range (min, max)") {
+    assertEquivalent('a > 300, 'a > 300.toShort)
+    assertEquivalent('a >= 500, 'a >= 500.toShort)
+    assertEquivalent('a === 32766, 'a === 32766.toShort)
+    assertEquivalent('a <=> 32766, 'a <=> 32766.toShort)
+    assertEquivalent('a <= -6000, 'a <= -6000.toShort)
+    assertEquivalent('a < -32767, 'a < -32767.toShort)
+  }
+
+  test("unwrap casts when cast is on rhs") {
+    val v = Short.MaxValue
+    assertEquivalent(Literal(v.toInt) < 'a, 'a.attr.falseIfNotNull)
+    assertEquivalent(Literal(v.toInt) <= 'a, Literal(v) === 'a)
+    assertEquivalent(Literal(v.toInt) === 'a, Literal(v) === 'a)
+    assertEquivalent(Literal(v.toInt) <=> 'a, Literal(v) === 'a)
+    assertEquivalent(Literal(v.toInt) >= 'a, 'a.attr.trueIfNotNull)
+    assertEquivalent(Literal(v.toInt) > 'a, 'a =!= v)
+
+    assertEquivalent(Literal(30) <= 'a, Literal(30.toShort) <= 'a)
+  }
+
+  test("unwrap cast should have no effect when input is not integral type") {
+    assertEquivalent('b > 42.0, Cast('b, DoubleType) > 42.0)
+    assertEquivalent('b >= 42.0, Cast('b, DoubleType) >= 42.0)
+    assertEquivalent('b === 42.0, Cast('b, DoubleType) === 42.0)
+    assertEquivalent('b <=> 42.0, Cast('b, DoubleType) <=> 42.0)
+    assertEquivalent('b <= 42.0, Cast('b, DoubleType) <= 42.0)
+    assertEquivalent('b < 42.0, Cast('b, DoubleType) < 42.0)
+    assertEquivalent(Literal(42.0) > 'b, Literal(42.0) > Cast('b, DoubleType))
+    assertEquivalent(Literal(42.0) >= 'b, Literal(42.0) >= Cast('b, DoubleType))
+    assertEquivalent(Literal(42.0) === 'b, Literal(42.0) === Cast('b, DoubleType))
+    assertEquivalent(Literal(42.0) <=> 'b, Literal(42.0) <=> Cast('b, DoubleType))
+    assertEquivalent(Literal(42.0) <= 'b, Literal(42.0) <= Cast('b, DoubleType))
+    assertEquivalent(Literal(42.0) < 'b, Literal(42.0) < Cast('b, DoubleType))
+  }
+
+  test("unwrap cast should skip when expression is non-deterministic") {
+    assertEquivalent(Cast(First('a, ignoreNulls = true), IntegerType) <=> positiveInt,
+      Cast(First('a, ignoreNulls = true), IntegerType) <=> positiveInt)
+    assertEquivalent(Cast(First('a, ignoreNulls = true), IntegerType) <=> negativeInt,
+      Cast(First('a, ignoreNulls = true), IntegerType) <=> negativeInt)
```

Comment:
The following will show the test intention more clearly.
```scala
-    assertEquivalent(Cast(First('a, ignoreNulls = true), IntegerType) <=> positiveInt,
-      Cast(First('a, ignoreNulls = true), IntegerType) <=> positiveInt)
-    assertEquivalent(Cast(First('a, ignoreNulls = true), IntegerType) <=> negativeInt,
-      Cast(First('a, ignoreNulls = true), IntegerType) <=> negativeInt)
+    Seq(positiveInt, negativeInt).foreach { v =>
+      val e = Cast(First('a, ignoreNulls = true), IntegerType) <=> v
+      assertEquivalent(e, e)
+    }
```

Reply:
👍